core/search/nested_filter.cpp

```diff
@@ -476,26 +496,39 @@ class MinMatcher : public Merger,
 };
 
 template<typename A, typename Visitor>
-auto ResolveMatchType(Match match, score_t none_boost, A&& aggregator,
-                      Visitor&& visitor) {
-  if (match == kMatchNone) {
-    return visitor(NoneMatcher{std::forward<A>(aggregator), none_boost});
-  } else if (match == kMatchAny) {
-    return visitor(AnyMatcher<A>{std::forward<A>(aggregator)});
-  } else if (match == kMatchAll) {
-    return visitor(AllMatcher<A>{std::forward<A>(aggregator)});
-  } else if (match.IsMinMatch()) {
-    assert(doc_limits::eof(match.Max));
-    return visitor(MinMatcher<A>{match.Min, std::forward<A>(aggregator)});
-  } else {
-    return visitor(RangeMatcher<A>{match, std::forward<A>(aggregator)});
-  }
+auto ResolveMatchType(const sub_reader& segment,
+                      const std::variant<Match, DocIteratorProvider>& match,
+                      score_t none_boost, A&& aggregator, Visitor&& visitor) {
+  return std::visit(
```

Comment:
Also for all visit in this pr, maybe better to use?
```
template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
```
https://en.cppreference.com/w/cpp/utility/variant/visit

Reply:
I anyway have to write all these implementation, don't see much difference TBH